### DISCUSSION TITLE
how to select and deselect a tool without using the mouse click?

### DISCUSSION THREAD

**Discussion by vladQll** [xournalpp/xournalpp#4631] [Q&A]
Hello to all friends, I have been using xournal++ for a long time and I think the software is great and I am very grateful to the programmers for maintaining it and I do not want to waste your time but I think that the solution that I am looking for can be very useful to other people that maybe like me use a graphic tablet (in my case it is a wacom-ctl4100).

the tablet has 4 buttons each one I have configured as a keyboard shortcut, and these shortcuts are linked to some functions that go through the different tools every time I press them and these functions are described in the main of the plugin ShorcutPlugin.

for example:
button1 = Alt + s = PEN , HIGHLIGHTER
button2 = Alt + q = Colors (plugin cycleColors)
button3 = Alt + d = RULER , TOOL_DRAW_ARROW, TOOL_DRAW_DOUBLE_ARROW, TOOL_DRAW_ELLIPSE, TOOL_DRAW_RECT
button4 = Alt + l = PLAIN, DASH, DASH_DOT, DOT
 
Well but I am going to focus more on the button3 that corresponds to the figures when I refer to select and deselect a tool, the steps to reproduce the action are:
in my case as it is assigned to an Alt-d , if I press once the straight line is selected , and if I press again the arrow is selected and so on , however for a manual reproduction would be to choose the tool with the mouse or with the pencil of the tablet and if you want to stop using it you have to click on the selected tool , is there where I have a little problem , as I use the button of the tablet to go through the different figures that I use , if I want to stop using I have to deselect manually with the mouse or the pencil.

So I would like to know what I can add to the function that I have that allows me to deselect the figure after using it. I have two ideas for the behavior but I would like to know how I could program that since I don't know how to program much:

1. after going through all the shapes, when I press the button again deselect the last shapes used.
2. another way that occurs to me is that the behavior is as follows I press the button for the first time, and the shapes is selected, I press again and the shapes is deselected, and when I press again it goes to the next shapes and if I press again it is deselected and if I press again it goes to the next figure and so on.

I am sending a reference video of the behavior I now have
https://user-images.githubusercontent.com/28522011/216484446-3ff1a150-d663-43b2-a6b4-2cbd7ba15e7f.mp4

this is the main where the functions where the tools are traversed are located:
[main.zip](https://github.com/xournalpp/xournalpp/files/10574573/main.zip)

**Comment by fengzyf** [comment]
Using a variable called callDrawingtypenumber can achieve both ideas.
> after going through all the shapes, when I press the button again deselect the last shapes used.
```
local callDrawingtypenumber = 0

function drawingtype()
  callDrawingtypenumber = callDrawingtypenumber % (#drawingtypeList + 1) + 1
  if callDrawingtypenumber ~= (#drawingtypeList + 1) then
    currentDrawingtype = currentDrawingtype % #drawingtypeList + 1
    app.uiAction({["action"] = "ACTION_" .. drawingtypeList[currentDrawingtype]})
    print("ACTION_" .. drawingtypeList[currentDrawingtype])
  else
    app.uiAction({["action"] = "ACTION_" .. drawingtypeList[currentDrawingtype], ["enabled"] = false})
  end
end
``` 

> another way that occurs to me is that the behavior is as follows I press the button for the first time, and the shapes is selected, I press again and the shapes is deselected, and when I press again it goes to the next shapes and if I press again it is deselected and if I press again it goes to the next figure and so on.

```
local callDrawingtypenumber = 0

function drawingtype()
  callDrawingtypenumber = callDrawingtypenumber % 2 + 1
  if callDrawingtypenumber == 1 then
    currentDrawingtype = currentDrawingtype % #drawingtypeList + 1
    app.uiAction({["action"] = "ACTION_" .. drawingtypeList[currentDrawingtype]})
    print("ACTION_" .. drawingtypeList[currentDrawingtype])
  else
    app.uiAction({["action"] = "ACTION_" .. drawingtypeList[currentDrawingtype], ["enabled"] = false})
  end
end
```
